### DISCUSSION TITLE
CORS configuration for Admin REST API - /admin/serverinfo

### DISCUSSION THREAD

**Discussion by ammartinCIC** [keycloak/keycloak#14598] [Q&A]
Hello,

Environment:
* Keycloak: 19.0.1
* Deployed in docker
* Traefik as proxy

I'm trying to make a request to `/admin/serverinfo` endpoint directly from my web page (Vue), but I get:

`Access to XMLHttpRequest at 'https://dev.xxx.com/auth/admin/serverinfo' from origin 'http://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource`

I know that I can configure CORS' headers in Realms Settings -> Security Defends but this request seems it's out of any realms because the url isn't `auth/realms/xxxx/serverinfo`.

The configuration on the master and my custom realm for Security Defends are:
![image](https://user-images.githubusercontent.com/19465895/192476727-29de7726-872c-40b6-a4a6-db868ad6e9f4.png)

But this headers doesn't appear in the response to the endpoint:
![image](https://user-images.githubusercontent.com/19465895/192477004-e9a7946b-1f75-4092-badb-28a87fd70889.png)

Is there any way to configure CORS' headers for Admin REST API? 

Thanks in advance,
Abel

**Comment by garronej** [comment]
Hello @ammartinCIC,  

Di you find a workaround?

**Comment by danielFesenmeyer** [comment]
Afaik, CORS settings can be configured per client, as "Web Origins": https://www.keycloak.org/docs/latest/server_admin/#con-basic-settings_server_administration_guide:~:text=for%20more%20information.-,Web%20Origins,-Enter%20a%20URL

**Comment by garronej** [comment]
Hello @danielFesenmeyer,  

Thanks for your answer! Yes the cors can indeed be configured by client.  
But I'm looking for enabling `/admin/serverinfo` to be queried from an other origin.  
Do you have any idea?

**Reply by danielFesenmeyer** [reply]
It looks like realm and client are resolved from the token, in case of this endpoint: https://github.com/keycloak/keycloak/blob/26a9516126c0dded4953681918f8b7b447e5b523/services/src/main/java/org/keycloak/services/resources/admin/AdminRoot.java#L252

**Reply by danielFesenmeyer** [reply]
Also note that the endpoint requires authorization. In case there is no authorization provided, the client can not be resolved and it's therefore not possible to retrieve the CORS settings for the client. Unauthorized will therefore always return an error for CORS requests.

**Comment by garronej** [comment]
@danielFesenmeyer, thank you so much for taking the time to help with this!

For context, this is the final blocker preventing me from shipping the Admin UI customization support with [Keycloakify](https://keycloakify.dev). So resolving this issue is high stake, I'll add you to the contributor list if you can help me figure this out 😄.

> It looks like realm and client are resolved from the token, in case of this endpoint:

I’m not entirely sure I understand what you mean here. What I need is the ability to run the Admin UI React SPA from localhost. I can add `Web Origins: *` to the `security-admin-console` client, but I still need to query the `/admin/serverinfo` endpoint. I’m not sure how to configure the policies for this endpoint.

> Also note that the endpoint requires authorization. In case there is no authorization provided, the client can’t be resolved, making it impossible to retrieve the CORS settings for the client. Unauthorized requests will always return a CORS error.

I agree, we’re on the same page about this.

**Comment by garronej** [comment]
Ok I see:  

https://github.com/keycloak/keycloak/blob/26a9516126c0dded4953681918f8b7b447e5b523/services/src/main/java/org/keycloak/services/resources/admin/AdminRoot.java#L273  

I was too anxious to answer. I'll investigate, that's an exellent pointer

**Comment by garronej** [comment]
@danielFesenmeyer Let's go 🚀 ,  

I was able to figure out from your pointer that I can add an hard codded claim to the AccessToken to make it work!  

<img width="1442" alt="image" src="https://github.com/user-attachments/assets/15109291-14bc-430b-9bd9-2d08275e153b" />


Thank you very very much

**Reply by NickT-78** [reply]
Hi,

Thank for sharing this solution, it really helped me solve the multi-realm with multi frontend URL issue.
Here are the extra toggle settings that work for me:
<img width="358" height="529" alt="image" src="https://github.com/user-attachments/assets/f5d24377-b9f3-4c51-aa7b-33e4fa1091cf" />
